test/integration/apiserver/crd_regression_test.go

```diff
@@ -0,0 +1,133 @@
+	apiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+	"k8s.io/utils/pointer"
+)
+
+// Regression test for https://issues.k8s.io/109099
+func TestCRDExponentialRecursionBug(t *testing.T) {
+	server, err := apiservertesting.StartTestServer(t, apiservertesting.NewDefaultTestServerOptions(), nil, framework.SharedEtcd())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.TearDownFn()
+	config := server.ClientConfig
+
+	apiExtensionClient, err := clientset.NewForConfig(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	crd := &apiextensionsv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "crontabs.stable.example.com",
+		},
+		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+			Group: "stable.example.com",
+			Scope: apiextensionsv1.ClusterScoped,
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Plural:   "crontabs",
+				Singular: "crontab",
+				Kind:     "CronTab",
+				ListKind: "CronTabList",
+			},
+			Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+				{
+					Name:    "v1beta1",
+					Served:  true,
+					Storage: true,
+					Schema: &apiextensionsv1.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+							XPreserveUnknownFields: pointer.BoolPtr(true),
+							Type:                   "object",
+							Properties:             map[string]apiextensionsv1.JSONSchemaProps{},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	crd, err = apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Create(context.TODO(), crd, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Failed to list custom resource: the server could not find the requested resource
+	err = wait.PollImmediate(100*time.Millisecond, 15*time.Second, func() (bool, error) {
+		_, err := apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), crd.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	gvr := schema.GroupVersionResource{
+		Group:    crd.Spec.Group,
+		Version:  crd.Spec.Versions[0].Name,
+		Resource: crd.Spec.Names.Plural,
+	}
+
+	crClient := dynamicClient.Resource(gvr)
+
+	instance := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": gvr.Group + "/" + gvr.Version,
+			"kind":       crd.Spec.Names.Kind,
+			"metadata": map[string]interface{}{
+				"name": "instanceame",
+			},
+			"spec": map[string]interface{}{},
+		},
+	}
+
+	var m map[string]interface{}
+	m = instance.Object["spec"].(map[string]interface{})
+	for i := 0; i < 50; i++ {
+		m[fmt.Sprintf("field%d", i)] = map[string]interface{}{}
+		m = m[fmt.Sprintf("field%d", i)].(map[string]interface{})
+	}
+	m["final"] = "final"
+
+	_, err = crClient.Create(context.TODO(), instance, metav1.CreateOptions{})
```

Comment:
why this flakes here with 
```
    crd_regression_test.go:131: Failed to create custom resource: the server could not find the requested resource
```

Reply:
Yeah, it looks like your poll isn't doing the right thing? Can you try polling the discovery instead? I'm assuming you've taken this from somewhere.

Reply:
right, I forget how this worked, it has to be present on discovery in order to be used, I fixed it now